### ISSUE TITLE
Composite primary keys

### ISSUE THREAD

**Issue by cmelchior** [realm/realm-java#1129]
Realm should also support composite primary keys. Reusing the ```@PrimaryKey``` annotation should be relatively simple.

```
public class CompositeKeyObj extends RealmObject {

  @PrimaryKey
  private int compositePartOne;

  @PrimaryKey
  private int compositePartTwo;
}
```

Primary keys are currently implemented in the binding. This should be implemented at the core level: https://github.com/realm/realm-core/issues/831

**Comment by senorcris** [comment]
This feature would be great is it still on the roadmap? If not are there any work arounds that could be used in the meantime to get composite keys?

**Comment by shawnwall** [comment]
+1

**Comment by ramanbranavitski** [comment]
How can I check status of this task?

**Comment by carmas123** [comment]
+1

**Comment by hunain-liaquat** [comment]
+1

**Comment by Saketme** [comment]
This would solve a problem I'm currently facing. Would be a great feature to have!

**Comment by dongseok0** [comment]
It's been 6 months. What's status?

**Comment by yuwu** [comment]
+1

**Comment by willf80** [comment]
+1

**Comment by fewspider** [comment]
+1

**Comment by yuwu** [comment]
+1

**Comment by pratamawijaya** [comment]
+1

**Comment by mksengun** [comment]
+1

**Comment by marchy** [comment]
+1

**Comment by mmartikainen** [comment]
+1

**Comment by marchy** [comment]
One workaround we've used as a way of getting around this limitation is to create an 'id' property which simply concatenates the different fields that you want to use in your composite key. This way you get the uniqueness integrity checks but you can still program against the proper abstractions (ie: foreign keys/objects that make up your composite key - especially useful for "link table" objects between your various entities.

**Comment by wuwu2000** [comment]
+1

**Comment by razelsoco** [comment]
+1

**Comment by mgdiez** [comment]
+1

**Comment by realpunk** [comment]
A workaround what I have been using is to create the unique id of combining two id's before assigning it to realm object eg.


```swift
class Pet : Object {
dynamic var combinedPrimary : String = ""
override class func primaryKey() -> String? {
      return "combinedPrimary" 
   }
}
```
```swift
let combinedKey = "\(person.id)\(pet.id)"
pet.combinedPrimary = combinedKey 
```

**Comment by marchy** [comment]
@realpunk yup I think that's the recommended workaround. It's probably why the team hasn't been prioritizing this very much - along with the fact that Realm doesn't really expose foreign keys in any way (I'm no expert here, just my experience with it). So unlike traditional ORMs where you want to keep foreign keys around on objects to link to them in Realm you link to the object instance directly and the backing ID/structure is hidden from the programming model.

For readability/sanity purposes I'd recommend throwing an extra ":" separator in there so you can re-infer the original IDs if needed:

ie: "\(person.id):\(pet.id)"

**Comment by StefanoGR** [comment]
+1

**Comment by tbeech** [comment]
+1

I'm not so sure that concatenating two keys together "really" solves the problem. Generally compound PK's are used so that you can query part of the key to return a list of related items. For instance I might want to log a set of actions where the log entry pk might be log.id + log.actionId as separate columns. Concatenating the two prevents me from querying against the db such that:

`
NSString *logId = @"SOME UUID";
NSPredicate *query = [NSPredicate predicateWithFormat=@"%K == %@", @"id", logId];
RLMResults<LogEntry *> *results = [LogEntry objectsWithPredicate: query];
`

Compound PK's also go hand in hand with unique constraints which Realm currently does not support.

**Comment by dlleixap** [comment]
+1

**Comment by kojadin** [comment]
+1

**Comment by cgathergood** [comment]
+1

**Comment by Zhuinden** [comment]
@tbeech no one is stopping you from also adding `@Index` annotated fields for parts of the handmade-composite primary key....

**Comment by alexxxdev** [comment]
+1

**Comment by DanteNguyen1008** [comment]
+1

**Comment by novinfard** [comment]
+1

**Comment by Zhuinden** [comment]
Guys, seriously, you can just concatenate the values together into a String.

**Comment by Dragas** [comment]
Or you know, instead of writing +1 just click that "Subscribe" button.

**Comment by Zhuinden** [comment]
@Dragas agreed. Also, just `+1` doesn't provide a use-case or anything. For me, concatenating values into a single indexed String field has always solved "compound primary key" problems.

**Comment by HsiangLeekwok** [comment]
+1

**Comment by marafat** [comment]
I am currently using something similar to the example by @HsiangLeekwok. But there are a couple of problems with that: 
* Since this `primaryKey` string is part of the realm object, it needs to have a public mutator. There is no clean way to convey to the rest of your team to avoid using this mutator and rather call `compoundPrimaryKey()`. I am currently annotating it as `@Deprecated` with a link ref to my version of `compoundPrimaryKey()`.
* You also have to make it clear that calling `compoundPrimaryKey` must not only be done on an unmanaged obj, it should also be on an obj that has both `id` & `type` none null. If both were to be marked as `@PrimaryKey`, it comes naturally knowing that you have to set primary keys before inserting objs into realm.

**Comment by Zhuinden** [comment]
> Since this primaryKey string is part of the realm object, it needs to have a public mutator.

Are you sure? I don't think it needs to have a public mutator since 0.88.0.

**Comment by marafat** [comment]
@Zhuinden I think you are right. I just tried with a simple test app where I didn't even generate a setter for the "compound primary key string". And was able to insert a new instance to the realm. I verified the obj by using this compound key to query for the obj and print it to logcat, and I also verified it with the realm browser.

In my defense, the [docs](https://realm.io/docs/java/latest/#models) mentions "Standard setters and getters", and they are exposed as `public` in the example.

**Comment by renattomachado** [comment]
+1

**Comment by PedroHFranco** [comment]
+1

**Comment by debjitk** [comment]
+1

**Comment by behelit** [comment]
Any clean code to do this? The suggestions so far seem horrible

**Comment by Zhuinden** [comment]
@behelit the current solution is still to keep your own "compound" String key that is constructed by concatenating whatever fields you wanted it to be based on

Once you create a managed object with a given primary key, you can't mutate its primary key though, so keep that in mind

**Comment by hmedat2008** [comment]
+18

**Comment by CsabaMiomni** [comment]
+1 million

**Comment by MartinaMuscarella** [comment]
+1

**Comment by svazquezcds** [comment]
+1

**Comment by GBenoitG** [comment]
+1

**Comment by lzbrady** [comment]
I am getting a list of object that I write to the realm that I would like to generate a compound key for. Do I have to loop over the list of incoming data and generate the compound key for each object? Or is there some method I can override or something to generate the primary key as it is being written to the realm?

**Comment by Zhuinden** [comment]
> Do I have to loop over the list of incoming data and generate the compound key for each object? 

yes

>  Or is there some method I can override or something to generate the primary key as it is being written to the realm?

no

**Comment by A-BEECHEY** [comment]
When you want to persist a json array directly into a realm database you CANNOT ".. just concatenate the values together into a String."

On the whole REALM is a brilliant mobile database, seems a pity that there a few "potholes" that have existed for some time which seem (on the face of it) to be obviously in need of fixing.

The other is an AUTO incrementing primary key.

Sure you can work round these, but is that really an acceptable solution in the 21st Century?

If Realm wants to be take seriously can then take the time to iron out these wrinkles?

**Comment by Zhuinden** [comment]
> When you want to persist a json array directly into a realm database you CANNOT ".. just concatenate the values together into a String."

I mean it worked for me, I just had to use a stream JSON parser to do it.

**Comment by gkalmpenis** [comment]
I am also in need of this, will use the workaround with string concatenation for now and just commenting in case someone has the capacity to implement it 😄
